runtime/src/accounts_db.rs

```diff
@@ -2886,59 +2929,152 @@ impl AccountsDB {
+                        None
+                    }
                 })
                 .unzip();
-            let aligned_total_size = self.page_align(total_size);
-
-            // This ensures that all updates are written to an AppendVec, before any
-            // updates to the index happen, so anybody that sees a real entry in the index,
-            // will be able to find the account in storage
-            let flushed_store =
-                self.create_and_insert_store(slot, aligned_total_size, "flush_slot_cache");
-            self.store_accounts_custom(
-                slot,
-                &accounts,
-                &hashes,
-                Some(Box::new(move |_, _| flushed_store.clone())),
-                None,
-                false,
-            );
-            // If the above sizing function is correct, just one AppendVec is enough to hold
-            // all the data for the slot
-            assert_eq!(
-                self.storage
-                    .get_slot_stores(slot)
-                    .unwrap()
-                    .read()
-                    .unwrap()
-                    .len(),
-                1
-            );
+
+            let is_dead_slot = accounts.is_empty();
+            // Remove the account index entries from earlier roots that are outdated by later roots.
+            // Safe because queries to the index will be reading updates from later roots.
+            self.purge_slot_cache_keys(slot, purged_slot_pubkeys, pubkey_to_slot_set, is_dead_slot);
+
+            if !is_dead_slot {
+                let aligned_total_size = self.page_align(total_size);
+                // This ensures that all updates are written to an AppendVec, before any
+                // updates to the index happen, so anybody that sees a real entry in the index,
+                // will be able to find the account in storage
+                let flushed_store =
+                    self.create_and_insert_store(slot, aligned_total_size, "flush_slot_cache");
+                self.store_accounts_custom(
+                    slot,
+                    &accounts,
+                    &hashes,
+                    Some(Box::new(move |_, _| flushed_store.clone())),
+                    None,
+                    false,
+                );
+                // If the above sizing function is correct, just one AppendVec is enough to hold
+                // all the data for the slot
+                assert_eq!(
+                    self.storage
+                        .get_slot_stores(slot)
+                        .unwrap()
+                        .read()
+                        .unwrap()
+                        .len(),
+                    1
+                );
+            }
 
             // Remove this slot from the cache, which will to AccountsDb readers should look like an
             // atomic switch from the cache to storage
             assert!(self.accounts_cache.remove_slot(slot).is_some());
         }
```

Comment:
if we can move the blow `if` as I said https://github.com/solana-labs/solana/pull/14596/files?diff=split&w=1#r560385136,
we can further simplify this `fn` by removing `slot_existed_in_cache`:

```rust
  ...
  true
} else {
  false
}
```